packages/block-editor/src/store/test/actions.js

```diff
@@ -595,12 +642,206 @@ describe( 'actions', () => {
+			} );
+
+			insertBlocks(
+				blocks,
+				5,
+				'testrootid',
+				false,
+				0
+			)( { select, dispatch } );
+
+			expect( updateInsertUsageSpy ).toHaveBeenCalledWith( blocks );
+		} );
+	} );
+
+	describe( 'updateInsertUsage', () => {
+		it( 'should record recently used blocks', () => {
+			const setPreference = jest.fn();
+			const registry = {
+				dispatch: () => ( {
+					set: setPreference,
+				} ),
+				select: () => ( {
+					get: () => {},
+					getActiveBlockVariation: () => {},
+				} ),
+			};
+
+			updateInsertUsage( [
+				{
+					clientId: 'bacon',
+					name: 'core/embed',
+				},
+			] )( { registry } );
+
+			expect( setPreference ).toHaveBeenCalledWith(
+				'core',
+				'insertUsage',
+				{
+					'core/embed': {
+						time: expect.any( Number ),
+						count: 1,
+						insert: { name: 'core/embed' },
+					},
+				}
+			);
+		} );
+
+		it( 'merges insert usage if more blocks are added of the same type', () => {
+			const setPreference = jest.fn();
+			const registry = {
+				dispatch: () => ( {
+					set: setPreference,
+				} ),
+				select: () => ( {
+					// simulate an existing embed block.
+					get: () => ( {
+						'core/embed': {
+							time: 123456,
+							count: 1,
+							insert: { name: 'core/embed' },
+						},
+					} ),
+					getActiveBlockVariation: () => {},
+				} ),
+			};
+
+			updateInsertUsage( [
+				{
+					clientId: 'eggs',
+					name: 'core/embed',
+				},
+				{
+					clientId: 'bacon',
+					name: 'core/block',
+					attributes: { ref: 123 },
+				},
+			] )( { registry } );
+
+			expect( setPreference ).toHaveBeenCalledWith(
+				'core',
+				'insertUsage',
+				{
+					// The reusable block has a special case where each ref is
+					// stored as though an individual block, and the ref is
+					// also recorded in the `insert` object.
+					'core/block/123': {
+						time: expect.any( Number ),
+						count: 1,
+						insert: { name: 'core/block', ref: 123 },
+					},
+					'core/embed': {
+						time: expect.any( Number ),
+						count: 2,
+						insert: { name: 'core/embed' },
+					},
+				}
+			);
+		} );
+
+		describe( 'block variations handling', () => {
+			const blockWithVariations = 'core/test-block-with-variations';
+
+			it( 'should return proper results with both found or not found block variation matches', () => {
+				const setPreference = jest.fn();
+				const registry = {
+					dispatch: () => ( {
+						set: setPreference,
+					} ),
+					select: () => ( {
+						get: () => {},
+						// simulate an active block variation:
+						// - 'apple' when the fruit attribute is 'apple'.
+						// - 'orange' when the fruit attribute is 'orange'.
+						getActiveBlockVariation: (
+							blockName,
+							{ fruit } = {}
+						) => {
+							if ( blockName === blockWithVariations ) {
+								if ( fruit === 'orange' )
+									return { name: 'orange' };
+								if ( fruit === 'apple' )
+									return { name: 'apple' };
+							}
```

Comment:
The added complexity to these tests is not great, but I don't really have any better suggestions for now 😅